Indexing/SubstitutionTree.hpp

```diff
@@ -678,71 +733,362 @@ class SubstitutionTree
+    LeafIterator(LeafIterator&&) = default;
+    LeafIterator& operator=(LeafIterator&&) = default;
+    DECL_ELEMENT_TYPE(Leaf*);
+    LeafIterator(SubstitutionTree* st);
     bool hasNext();
-    Leaf* next()
-    {
-      ASS(_curr->isLeaf());
-      return static_cast<Leaf*>(_curr);
-    }
+    Leaf* next();
   private:
-    Node** _nextRootPtr;
-    Node** _afterLastRootPtr;
+    void skipToNextLeaf();
     Node* _curr;
     Stack<NodeIterator> _nodeIterators;
   };
 
-  typedef pair<pair<LeafData*, ResultSubstitutionSP>,UnificationConstraintStackSP> QueryResult;
 
 
-  class GenMatcher;
+   /**
+   * Class that supports matching operations required by
+   * retrieval of generalizations in substitution trees.
+   */
+  class GenMatcher
+  {
+    static unsigned weight(Literal* l) { return l->weight(); }
+    static unsigned weight(TermList t) { return  t.weight(); }
+  public:
+    GenMatcher(GenMatcher&&) = default;
+    GenMatcher& operator=(GenMatcher&&) = default;
+
+    /**
+     * @b nextSpecVar Number higher than any special variable present in the tree.
+     * 	It's used to determine size of the array that stores bindings of
+     * 	special variables.
+     */
+    template<class TermOrLit>
+    GenMatcher(TermOrLit query, unsigned nextSpecVar)
+      : _boundVars()
+      , _specVars()
+      , _maxVar(weight(query) - 1)
+      , _bindings()
+    {
+      if(_specVars->size()<nextSpecVar) {
+        //_specVars can get really big, but it was introduced instead of hash table
+        //during optimizations, as it raised performance by abour 5%.
+        _specVars->ensure(max(static_cast<unsigned>(_specVars->size()*2), nextSpecVar));
+      }
+      _bindings->ensure(weight(query));
+    }
+
+
+
+    CLASS_NAME(SubstitutionTree::GenMatcher);
+    USE_ALLOCATOR(GenMatcher);
+
+    /**
+     * Bind special variable @b var to @b term. This method
+     * should be called only before any calls to @b matchNext()
+     * and @b backtrack().
+     */
+    void bindSpecialVar(unsigned var, TermList term)
+    {
+      (*_specVars)[var]=term;
+    }
+    /**
+     * Return term bound to special variable @b specVar
+     */
+    TermList getSpecVarBinding(unsigned specVar)
+    { return (*_specVars)[specVar]; }
+
+    bool matchNext(unsigned specVar, TermList nodeTerm, bool separate=true);
+    bool matchNextAux(TermList queryTerm, TermList nodeTerm, bool separate=true);
+    void backtrack();
+    bool tryBacktrack();
+
+    ResultSubstitutionSP getSubstitution(Renaming* resultNormalizer);
+
+    int getBSCnt()
+    {
+      int res=0;
+      VarStack::Iterator vsit(*_boundVars);
+      while(vsit.hasNext()) {
+    if(vsit.next()==BACKTRACK_SEPARATOR) {
+      res++;
+    }
+      }
+      return res;
+    }
+
+  protected:
+    static const unsigned BACKTRACK_SEPARATOR=0xFFFFFFFF;
+
+    struct Binder;
+    struct Applicator;
+    class Substitution;
+
+    Recycled<VarStack> _boundVars;
+    Recycled<DArray<TermList>, NoReset> _specVars;
```

Comment:
Why `NoReset`?

Reply:
The reason for this is that we do not need to, as we override all values that will be read anyways. Therefore we will safe time by not resetting. This behaviour was there before, I just implemented it using `Recycled`.

Reply:
*I added a comment